### PR TITLE
 netifd: change RPS/XPS handling to all CPUs and disable by default 

### PR DESCRIPTION
**Pull request by alanswanson**
netifd: change RPS/XPS handling to all CPUs and disable by default

The current implementation is significantly lowering lantiq performace [1][2] by using RPS with non-irq CPUs and XPS with alternating CPUs.

The previous netifd implementation (by default but could be configured) simply used all CPUs and this patch essentially reverts to this behaviour.

The only document suggesting using non-interrupt CPUs is Red Hat [3] where if the network interrupt rate is extremely high excluding the CPU that handles network interrupts *may* also improve performance.

The original packet steering patches [4] advise that optimal settings for the CPU mask seems to depend on architectures and cache hierarchy so one size does not fit all. It also advises that the overhead in processing for a lightly loaded server can cause performance degradation.

Ideally, proper IRQ balancing is a better option with the irqbalance daemon or by manually setting.

The kernel does not enable packet steering by default, so also disable in OpenWRT by default. (Though mvebu with its hardware scheduling issues [5] might want to enable packet steering by default.)

Change undocumented "default_ps" parameter to clearer "packet_steering" parameter. The old parameter was only ever set in target/linux/mediatek/base-files/etc/uci-defaults/99-net-ps and matched the default.

Rename the script to be more obvious that this is for packet steering only.

[1] https://forum.openwrt.org/t/18-06-4-speed-fix-for-bt-homehub-5a
[2] https://openwrt.ebilan.co.uk/viewtopic.php?f=7&t=1105
[3] https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/performance_tuning_guide/network-rps
[4] https://marc.info/?l=linux-netdev&m=125792239522685&w=2
[5] https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=2e1f6f1682d3974d8ea52310e460f1bbe470390f

### CODE REVIEW

**Comment by jow-** [comment]
LGTM from my side. Ping @blogic - he fought with this too afair.

**Comment by ahmar20** [comment]
If this gets added to OpenWrt then how this will be enabled because it says disabled by default?

**Comment by alanswanson** [comment]
As you can see in the patch, it remains at the same uci subsystem and has same bool activation as the old **_default_ps_** but now named **_packet_steering_** and defaults to off.  Since default_ps was undocumented I assume nobody knew about unless they read 20-smp-tune.sh or the old netifd code.

$ uci set network.globals.packet_steering=1

**Comment by ahmar20** [comment]
AFAIK there's only one setting in Global section in Interfaces and that is IPv6-Prefix of some sort but I hope this gets added there as well in LuCI otherwise it would still be remain unknown to everyone. Even though it's disabled by default the setting should be visible at least to make it noticeable. I understand this patch is not about the LuCI setting but this is just a thought.

**Comment by jeolives** [comment]
There used to be a setting [1][2] prior to the changes in netifd that necessitated this PR. It used to be called `network.globals.default_ps`.

Though the kernel doesn't enable RPS by default, a quick look through the repo for `CONFIG_RPS=y` and `CONFIG_XPS=y` shows quite a number of targets having it built in. I'm not a dev, but I still think it would be best to have it opt out. More advanced users will just disable it if they need to.

On a tangent with this PR, does this also make 7af1fb9faafbc842fc727c49108f5fc4edc08601 relevant again? Though I can see through the commit history that it causes problems on x86...


[1] https://github.com/MarvellEmbeddedProcessors/openwrt-dd/wiki/Working-with-network-interfaces
[2] http://wiki.espressobin.net/tiki-index.php?page=OpenWrt+-+initial+network+configuration

**Comment by mkresin** [comment]
You need to apply the option rename to `target/linux/mediatek/base-files/etc/uci-defaults/99-net-ps` as well:

```
target/linux/mediatek/base-files/etc/uci-defaults/99-net-ps:4:uci set network.globals.default_ps=1
```

File: package/network/config/netifd/files/etc/hotplug.d/net/20-smp-tune [hunk 1]
```diff
@@ -34,8 +34,8 @@ set_hex_val() {
 	echo "$val" > "$file"
 }
 
-default_ps="$(uci get "network.@globals[0].default_ps")"
-[ -n "$default_ps" -a "$default_ps" != 1 ] && exit 0
+packet_steering="$(uci get "network.@globals[0].packet_steering")"
+[ "$packet_steering" != 1 ] && exit 0
 
 exec 512>/var/lock/smp_tune.lock
```
Comment (ahmar20) [inline package/network/config/netifd/files/etc/hotplug.d/net/20-smp-tune]: Should the lock file not be renamed as well?

**Comment by alanswanson** [comment]
@ahmar20 Will submit a PR to luci to add a packet steering option after this is merged.

@jeolives Those two examples specifically disable packet steering. IRQ balancing is preferred, while packet steering is the advanced but not always better option. The "old" uci option has already been stated in the comments and code. RPS/XPS are compiled by default for smp kernels and cannot be normally be chosen in a normal kernel (menu)config and also cannot be a module so would always need to built in.

**Comment by mx2345** [comment]
Have been using this patch with my FRITZ3370 for a week now and it works fine.

I set the default_ps parameter to 0 a long time ago because I noticed that packet steering lowered performance on this device by quite a bit.

**Comment by blogic** [comment]
LGTM

**Comment by paravz** [comment]
packet_steering = 1 means it's disabled, which is counter intuitive. Since you creating a new variable you can flip the meaning:

packet_steering = 0 - disable rps/xps

**Comment by alanswanson** [comment]
@paravz No, follow the codes logic;
_[ "$packet_steering" != 1 ] && exit 0_

**Comment by paravz** [comment]
my bad, sorry, misread "!=" as "="

**Comment by pstlr78** [comment]
It is great to hear that this patch is tested and working. Is it then a fair assumption to say that it will be released in the next stable openwrt service release (i.e. 19.07.2)?

**Comment by paravz** [comment]
`[ "$packet_steering" != 1 ] && echo exit` - will disable packet steering by default for all boards other than mediatek, since this PR has "uci set network.globals.packet_steering=1" - perhaps the default uci settins should be changed?

This works great on ramips/mt7621(2 cores, 4 threads): disabling xps/rps significantly **reduces** cpu utilization while NATing about 1Gbps using TCP iperf (with hw offloads enabled).
Note significantly higher cpu/sy (system cpu usage %), system/in (# of interrupts) and system/cs (# of context switches)

1. xps disabled
```
# echo 0 >/sys/class/net/eth0/queues/rx-0/rps_cpus; echo 0 >/sys/class/net/eth0/queues/tx-0/xps_cpus

# grep . /sys/class/net/*/queues/*/?ps_cpus|grep eth
/sys/class/net/eth0.1/queues/rx-0/rps_cpus:0
/sys/class/net/eth0.1/queues/tx-0/xps_cpus:0
/sys/class/net/eth0.2/queues/rx-0/rps_cpus:0
/sys/class/net/eth0.2/queues/tx-0/xps_cpus:0
/sys/class/net/eth0/queues/rx-0/rps_cpus:0
/sys/class/net/eth0/queues/tx-0/xps_cpus:0

# vmstat -t -w 5
procs -----------------------memory---------------------- ---swap-- -----io---- -system-- --------cpu-------- -----timestamp-----
 r  b         swpd         free         buff        cache   si   so    bi    bo   in   cs  us  sy  id  wa  st                 
 1  0            0       466656         2756        15728    0    0     0     0 1911  141   1  25  74   0   0 04:55:24
 1  0            0       466484         2756        15728    0    0     0     0 1609  150   1  25  74   0   0 04:55:29
 1  0            0       466484         2756        15728    0    0     0     0  436  133   1  25  74   0   0 04:55:34
 1  0            0       466632         2756        15728    0    0     0     0  750  155   1  25  74   0   0 04:55:39
 2  0            0       466504         2756        15728    0    0     0     0  436  141   1  25  74   0   0 04:55:44
 1  0            0       466512         2756        15728    0    0     0     0  570  156   1  25  74   0   0 04:55:49
 1  0            0       466472         2756        15728    0    0     0     0  594  148   1  25  74   0   0 04:55:54
 1  0            0       466632         2756        15728    0    0     0     0 1880  154   1  25  74   0   0 04:55:59
 1  0            0       466592         2756        15728    0    0     0     0  760  138   1  25  74   0   0 04:56:04
 1  0            0       466712         2756        15728    0    0     0     0 2071  153   1  25  74   0   0 04:56:09
```

2. xpc enabled
```
# echo e >/sys/class/net/eth0/queues/rx-0/rps_cpus; echo 2 >/sys/class/net/eth0/queues/tx-0/xps_cpus

# grep . /sys/class/net/*/queues/*/?ps_cpus|grep eth
/sys/class/net/eth0.1/queues/rx-0/rps_cpus:0
/sys/class/net/eth0.1/queues/tx-0/xps_cpus:0
/sys/class/net/eth0.2/queues/rx-0/rps_cpus:0
/sys/class/net/eth0.2/queues/tx-0/xps_cpus:0
/sys/class/net/eth0/queues/rx-0/rps_cpus:e
/sys/class/net/eth0/queues/tx-0/xps_cpus:2

# vmstat -t -w 5
procs -----------------------memory---------------------- ---swap-- -----io---- -system-- --------cpu-------- -----timestamp-----
 r  b         swpd         free         buff        cache   si   so    bi    bo   in   cs  us  sy  id  wa  st                 
 2  0            0       466224         2756        15728    0    0     0     0 27820 1057   1  53  46   0   0 04:53:58
 2  0            0       466460         2756        15728    0    0     0     0 31332 1488   1  52  47   0   0 04:54:03
 1  0            0       466128         2756        15728    0    0     0     0 27950 1069   1  53  46   0   0 04:54:08
 2  0            0       466612         2756        15728    0    0     0     0 31986 1567   1  52  47   0   0 04:54:13
 2  0            0       466028         2756        15728    0    0     0     0 27532  903   1  53  46   0   0 04:54:18
 2  0            0       466468         2756        15728    0    0     0     0 31048 1435   1  53  46   0   0 04:54:23
 2  0            0       466580         2756        15728    0    0     0     0 31184 1594   1  52  47   0   0 04:54:28
 1  0            0       466648         2756        15728    0    0     0     0 31542 1380   1  53  46   0   0 04:54:33
 2  0            0       466304         2756        15728    0    0     0     0 30513 1420   1  50  50   0   0 04:54:38
 1  0            0       466700         2756        15728    0    0     0     0 31200 1460   1  52  47   0   0 04:54:43
```

**Comment by alanswanson** [comment]
@paravz It was requested by @mkresin in [comment](https://github.com/openwrt/openwrt/pull/2553#issuecomment-554639883) to update the mediatek configuration but keep it enabled per its current setting. If a developer does want it changed/disabled for mediatek I can update the PR.

**Comment by mkresin** [comment]
merged, thanks!

**Comment by pachulo** [comment]
Thanks for this @alanswanson !
@mkresin one question: will it be backported to 19.07?

**Comment by mkresin** [comment]
> @mkresin one question: will it be backported to 19.07?

Unlikely, as it isn't rather a performance fix than a bug fix.

**Pull request closed without merging**